File: base/src/test/java/io/spine/code/proto/TypeSetTest.java
/*
 * Copyright 2019, TeamDev. All rights reserved.
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.code.proto;

import com.google.common.collect.ImmutableCollection;
import com.google.common.collect.ImmutableSet;
import com.google.protobuf.DescriptorProtos.FileDescriptorSet;
import com.google.protobuf.Descriptors.FileDescriptor;
import io.spine.test.code.proto.MessageDecl;
import io.spine.type.MessageType;
import io.spine.type.TypeName;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static com.google.common.truth.Truth.assertThat;

@DisplayName("TypeSet should")
class TypeSetTest {

    private static final FileSet fileSet = FileSet.load();

    @Test
    @DisplayName("obtain messages and enums from a file")
    void fromFile() {
        @SuppressWarnings("OptionalGetWithoutIsPresent") /* The file is present in resources. */
        FileDescriptor file = fileSet.tryFind(FileName.of("google/protobuf/descriptor.proto"))
                                     .get();
        TypeSet typeSet = TypeSet.from(file);
        assertNotEmpty(typeSet);
        assertThat(typeSet.contains(TypeName.from(FileDescriptorSet.getDescriptor())))
                .isTrue();
    }

    @Test
    @DisplayName("obtain message and enums")
    void fromSet() {
        TypeSet typeSet = TypeSet.from(fileSet);
        assertNotEmpty(typeSet);
        // We have a number of test service declarations for testing annotations.
        assertThat(typeSet.serviceTypes())
                .isNotEmpty();
    }

    @Test
    @DisplayName("obtain top-level messages of the file")
    void topLevelMessages() {
        ImmutableSet<FileName> fileNames =
                ImmutableSet.of(FileName.of("spine/test/code/proto/type_set_test.proto"));
        FileSet set = fileSet.find(fileNames);
        ImmutableCollection<MessageType> types = TypeSet.topLevelMessages(set);
        assertThat(types).hasSize(1);

        MessageType onlyElement = types.asList()
                                       .get(0);
        assertThat(onlyElement.javaClass()).isEqualTo(MessageDecl.class);
    }

    void assertNotEmpty(TypeSet typeSet) {
        assertThat(typeSet.isEmpty())
                .isFalse();
        assertThat(typeSet.allTypes())
                .isNotEmpty();
        assertThat(typeSet.messageTypes())
                .isNotEmpty();
        assertThat(typeSet.enumTypes())
                .isNotEmpty();
    }
}

File: base/src/test/java/io/spine/code/proto/TypeSetTest.kt
/*
 * Copyright 2022, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package io.spine.code.proto

import com.google.protobuf.Any
import com.google.protobuf.DescriptorProtos.FileDescriptorSet
import com.google.protobuf.Duration
import io.kotest.matchers.collections.containOnly
import io.kotest.matchers.collections.shouldContain
import io.kotest.matchers.collections.shouldNotBeEmpty
import io.kotest.matchers.should
import io.kotest.matchers.shouldBe
import io.spine.type.MessageType
import io.spine.type.TypeName
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("`TypeSet` should")
internal class TypeSetTest {

    @Test
    fun `obtain messages and enums from a file`() {
        val fileName = FileName.of("google/protobuf/descriptor.proto")
        /* The file is present in resources. */
        val file = fileSet.tryFind(fileName).get()
        val typeSet = TypeSet.from(file)

        assertNotEmpty(typeSet)
        val expectedTypeName = TypeName.from(FileDescriptorSet.getDescriptor())
        typeSet.contains(expectedTypeName) shouldBe true
    }

    @Test
    @DisplayName("obtain message and enums")
    fun fromSet() {
        val typeSet = TypeSet.from(fileSet)
        assertNotEmpty(typeSet)

        // We have a number of test service declarations for testing annotations.
        typeSet.serviceTypes().shouldNotBeEmpty()
    }

    @Test
    fun `obtain message types from a 'TypeSet'`() {
        val messageTypes = TypeSet.onlyMessages(fileSet)

        messageTypes shouldContain MessageType.of(Any.getDefaultInstance())
        messageTypes shouldContain MessageType.of(Duration.getDefaultInstance())
    }

    @Test
    fun `obtain message types from a file descriptor`() {
        val fileName = FileName.of("google/protobuf/any.proto")
        val file = fileSet.tryFind(fileName).get()
        val messageTypes = TypeSet.onlyMessages(file)

        messageTypes should containOnly(MessageType.of(Any.getDefaultInstance()))
    }

    private fun assertNotEmpty(typeSet: TypeSet) {
        typeSet.isEmpty shouldBe false
        typeSet.allTypes().shouldNotBeEmpty()
        typeSet.messageTypes().shouldNotBeEmpty()
        typeSet.enumTypes().shouldNotBeEmpty()
    }

    companion object {
        private val fileSet = FileSet.load()
    }
}
